amlb/singularity.py

```diff
@@ -0,0 +1,238 @@
+that are preconfigured with a given automl framework, and that can be used to run a benchmark anywhere.
+The Singularity image embeds a version of the automlbenchmark app so that tasks are later run in local mode inside singularity,
+providing the same parameters and features allowing to import config and export results through mounted folders.
+The image is pulled form an existing docker, yet executed in singularity framework
+"""
+import logging
+import os
+import re
+
+from .container import ContainerBenchmark
+from .resources import config as rconfig, get as rget
+from .utils import dir_of, run_cmd
+
+
+log = logging.getLogger(__name__)
+
+
+class SingularityBenchmark(ContainerBenchmark):
+    """SingularityBenchmark
+    an extension of ContainerBenchmark to run benchmarks inside Singularity.
+    """
+
+    def __init__(self, framework_name, benchmark_name, constraint_name):
+        """
+
+        :param framework_name:
+        :param benchmark_name:
+        :param constraint_name:
+        """
+        super().__init__(framework_name, benchmark_name, constraint_name)
+        self._custom_image_name = rconfig().singularity.image
+        self.minimize_instances = rconfig().singularity.minimize_instances
+        self.container_name = 'singularity'
+        self.force_branch = rconfig().singularity.force_branch
+        self.custom_commands = self.framework_module.singularity_commands(
+            self.framework_def.setup_args,
+            setup_cmd=self.framework_def._setup_cmd
+        ) if hasattr(self.framework_module, 'singularity_commands') else ""
+
+    def _container_image_name(self, branch=None, return_docker_name=False):
+        """
+        Singularity Images would be located on the framework directory
+        We prefer to pull from docker, so we have to mind the docker tag
+        When downloading from Docker, the colon is changed to underscore
+        """
+        di = self.framework_def.image
+
+        # If we want to pull from docker, the separator is a colon for tag
+        separator = '_' if not return_docker_name else ':'
+        # Also, no need for author in image name
+        author = '' if not return_docker_name else f"{di.author}/"
+
+        if branch is None:
+            branch = rget().project_info.branch
+
+        image_name = "{author}{image}{separator}{tag}".format(
+            author=author,
+            separator=separator,
+            image=di.image if di.image else self.framework_def.name.lower(),
+            tag=re.sub(r"([^\w.-])", '.',
+                       '-'.join([di.tag if di.tag else self.framework_def.version.lower(), branch]))
+        )
+
+        # Make sure image is in the framework directory
+        if return_docker_name:
+            return image_name
+        else:
+            return os.path.join(
+            self._framework_dir,
+            image_name + '.sif'
+        )
```

Comment:
confusing indentation
```suggestion
            return os.path.join(self._framework_dir, image_name + '.sif')
```